src/kubelet/kubelet_test.go

```diff
@@ -85,19 +79,99 @@ func TestScraper(t *testing.T) {
 	}
 }
 
-func ExcludeMissingMetricsPendingPod(_ string, spec *definition.Spec, ent *integration.Entity) bool {
-	metricsToExcludeForPendingPods := []string{"memoryUsedBytes", "memoryWorkingSetBytes", "cpuUsedCores",
+// kubeletExclusions is a helper that returns all the exclusions needed to assert the kubelet metrics without getting
+// false negatives.
+func kubeletExclusions() []exclude.Func {
+	// Network metrics are known to be missing on some environments.
+	networkMetrics := []string{"net.rxBytesPerSecond", "net.txBytesPerSecond", "net.errorsPerSecond"}
+
+	// TODO: Unclear why we need to exclude node utilization metrics.
+	nodeUtilizationMetrics := []string{"allocatableCpuCoresUtilization", "allocatableMemoryUtilization"}
+
+	// Pods and containers that are not in a running state will not have these metrics.
+	notRunningMetrics := []string{"memoryUsedBytes", "memoryWorkingSetBytes", "cpuUsedCores",
 		"fsAvailableBytes", "fsCapacityBytes", "fsUsedBytes", "fsUsedPercent", "fsInodesFree", "fsInodes",
-		"fsInodesUsed", "containerID", "containerImageID", "isReady"}
-
-	for _, metricSet := range ent.Metrics {
-		if metricSet.Metrics["status"] == "Waiting" {
-			for _, m := range metricsToExcludeForPendingPods {
-				if m == spec.Name {
-					return true
-				}
-			}
-		}
+		"fsInodesUsed", "containerMemoryMappedFileBytes", "containerID", "containerImageID", "isReady", "podIP"}
+
+	// Utilization metrics will not be present if the corresponding limit/request is not present.
+	utilizationDependencies := map[string][]string{
+		"cpuLimitCores":        {"cpuCoresUtilization"},
+		"cpuRequestedCores":    {"requestedCpuCoresUtilization"},
+		"memoryLimitBytes":     {"memoryUtilization"},
+		"memoryRequestedBytes": {"requestedMemoryUtilization"},
+	}
+
+	// Regex to match limits/requests for CPU and Memory.
+	limitsRequestsRegex := regexp.MustCompile("(Limit|Requested)(Cores|Bytes)$")
+
+	return []exclude.Func{
+		// Network metrics
+		exclude.Metrics(networkMetrics...),
+
+		// Node utilization  metrics.
+		exclude.Exclude(exclude.Groups("node"), exclude.Metrics(nodeUtilizationMetrics...)),
+
+		// Exclude limits/requested metrics for nodes, pods and containers
+		exclude.Exclude(
+			exclude.Groups("node", "pod", "container"),
+			func(group string, spec *definition.Spec, e *integration.Entity) bool {
+				return limitsRequestsRegex.MatchString(spec.Name)
+			},
+		),
+		// Exclude metrics that depend on limits when those limits are not set.
+		exclude.Exclude(exclude.Groups("pod", "container"), exclude.Dependent(utilizationDependencies)),
```

Comment:
why excluding whole groups with ".Dependent" and not specific metrics?

Reply:
Ahhh I got it, It exclude everything from the group having those dependencies.. smart 😄